### PR TITLE
Desmos Rich Presence

### PR DESCRIPTION
**Pull request by Anshroid**
Desmos is an online graphing calculator with many other functions, such as a geometry designer and a scientific calculator: desmos.com. This presence is designed to provide information about all functions of the app.

**Screenshots of the Presence in operation:**

### CODE REVIEW

File: websites/D/Desmos/dist/metadata.json [hunk 1]
```diff
@@ -0,0 +1,22 @@
+{
+  "$schema": "https://schemas.premid.app/metadata/1.3",
```
Comment (EncryptedDev) [inline websites/D/Desmos/dist/metadata.json:2]:
```suggestion
  "$schema": "https://schemas.premid.app/metadata/1.4",
```

File: websites/D/Desmos/dist/metadata.json [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  "service": "Desmos",
+  "author": {
+    "name": "Anshroid",
+    "id": "434312233023963137"
+  },
+  "category": "other",
+  "logo": "https://i.imgur.com/MVpfVss.jpg",
+  "thumbnail": "https://i.imgur.com/XzSfNJ3.png",
```
Comment (EncryptedDev) [inline websites/D/Desmos/dist/metadata.json:10]: Thumbnail should be a promotional banner or screenshot of the homepage

File: websites/D/Desmos/presence.ts [hunk 3]
```diff
@@ -0,0 +1,167 @@
+const desmosPresence = new Presence({
+	clientId: "886254740399349852" //The client ID of the Application created at https://discordapp.com/developers/applications
+  }),
+  /*strings = desmosPresence.getStrings({
+	play: "presence.playback.playing",
+	pause: "presence.playback.paused"
+	//You can use this to get translated strings in their browser language
+  });*/
```
Comment (EncryptedDev) [inline websites/D/Desmos/presence.ts]: Remove comments 
```suggestion
	clientId: "886254740399349852"
  }),
```

File: websites/D/Desmos/presence.ts [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+  
+var url: string
+var graphing: number = 0
+var pageType: string
+var title: string = ""
+var numEquations: number = 0
```
Comment (EncryptedDev) [inline websites/D/Desmos/presence.ts:9]: Why are you assigning this all to 0 and empty strings?
```suggestion
let url: string, graphing: number, pageType: string, title: string, numEquations: number
```

File: websites/D/Desmos/presence.ts [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+const startTime: number = Date.now()
+  strings = desmosPresence.getStrings({
+    play: "presence.playback.playing",
+    pause: "presence.playback.paused"
+    //You can use this to get translated strings in their browser language
```
Comment (EncryptedDev) [inline websites/D/Desmos/presence.ts:14]:
```suggestion
```

File: websites/D/Desmos/presence.ts [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+  });
+
+let url: string,
+ graphing = 0,
+ pageType: string,
+ title = "",
+ numEquations = 0,
```
Comment (EncryptedDev) [inline websites/D/Desmos/presence.ts]: What's the point of reassigning them again? 
```suggestion
```

File: websites/D/Desmos/presence.ts [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+ startTime: number = Date.now();
```
Comment (EncryptedDev) [inline websites/D/Desmos/presence.ts]: startTime never changes, using `const` is more conventional
```suggestion
 const startTime: number = Date.now();
```

File: websites/D/Desmos/presence.ts [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+
+desmosPresence.on("UpdateData", async () => {
+  /*UpdateData is always firing, and therefore should be used as your refresh cycle, or `tick`. This is called several times a second where possible.
+
+	It is recommended to set up another function outside of this event function which will change variable values and do the heavy lifting if you call data from an API.*/
+	
+  // About the comment above: Is this really recommended? It appears to be not allowed according to the Presence Rules
+
```
Comment (EncryptedDev) [inline websites/D/Desmos/presence.ts:23]: Remove comments
```suggestion
```

File: websites/D/Desmos/presence.ts [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+desmosPresence.on("UpdateData", async () => {
+  /*UpdateData is always firing, and therefore should be used as your refresh cycle, or `tick`. This is called several times a second where possible.
+
+	It is recommended to set up another function outside of this event function which will change variable values and do the heavy lifting if you call data from an API.*/
+	
+  // About the comment above: Is this really recommended? It appears to be not allowed according to the Presence Rules
+  
+  url = document.URL;
+  if (url[url.length-1] === "/") {
+	url = url.substring(0, url.length-1)
```
Comment (EncryptedDev) [inline websites/D/Desmos/presence.ts:26]: Fix indentation

File: websites/D/Desmos/presence.ts [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+  }
+  var splitUrl = url.split("/")
+  var urlPage = splitUrl[splitUrl.length-1]
```
Comment (EncryptedDev) [inline websites/D/Desmos/presence.ts:29]:
```suggestion
  let splitUrl = url.split("/"),
   urlPage = splitUrl[splitUrl.length-1];
```

File: websites/D/Desmos/presence.ts [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+  if (urlPage === "www.desmos.com") { 
+	pageType = "Home Page" 
+  } else { 
+	pageType = urlPage[0].toUpperCase() + urlPage.substring(1) 
+  }
```
Comment (EncryptedDev) [inline websites/D/Desmos/presence.ts:34]: Fix indenation here, but curlies are not needed as you're just doing one thing
```suggestion
  if (urlPage === "www.desmos.com") pageType = "Home Page" 
  else  pageType = urlPage[0].toUpperCase() + urlPage.substring(1) 
```

File: websites/D/Desmos/presence.ts [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+  if (url[url.length - 1] == "/") 
+  	url = url.substring(0, url.length - 1);
+  
+  const splitUrl = url.split("/"),
+   urlPage = splitUrl[splitUrl.length - 1];
```
Comment (EncryptedDev) [inline websites/D/Desmos/presence.ts:39]: Why have you used the same variable name many times?

File: websites/D/Desmos/presence.ts [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+  if (urlPage == "www.desmos.com")  
+  	pageType = "Home Page"; 
+   else  
+  	pageType = urlPage[0].toUpperCase() + urlPage.substring(1);
```
Comment (EncryptedDev) [inline websites/D/Desmos/presence.ts:43]: I'm getting some crazy deja vu here, why is this the same as L42 - L46

File: websites/D/Desmos/presence.ts [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+   urlPage = splitUrl[splitUrl.length - 1];
+  if (urlPage == "www.desmos.com")  
+  	pageType = "Home Page"; 
+   else  
+  	pageType = urlPage[0].toUpperCase() + urlPage.substring(1); 
+  
+  if (url.includes("/calculator")) {
+	graphing = 2;	
+	title = document.getElementsByClassName("dcg-variable-title")[0].innerHTML;
+	numEquations = document.getElementsByClassName("dcg-template-expressioneach")[0].childElementCount;
```
Comment (EncryptedDev) [inline websites/D/Desmos/presence.ts:48]: Fix indentation

File: websites/D/Desmos/presence.ts [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+  } else if (url.includes("/geometry")) {
+	graphing = 1;	
+	title = document.getElementsByClassName("dcg-variable-title")[0].innerHTML;
+	pageType = "Geometry"
+	numEquations = 0
```
Comment (EncryptedDev) [inline websites/D/Desmos/presence.ts:53]: Fix indentation

File: websites/D/Desmos/presence.ts [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+	if (pageType === "Scientific" || pageType === "Fourfunction") {
+		numEquations = document.getElementsByClassName("dcg-basic-list")[0].childElementCount
+	} else if (pageType === "Matrix") {
+		numEquations = document.getElementsByClassName("dcg-matrix-list")[0].childElementCount
+	} else {
+		numEquations = 0
+	}
+  } else {
+	graphing = 0;
+  }
```
Comment (EncryptedDev) [inline websites/D/Desmos/presence.ts:66]: Remove curly brackets here where not appropriate

File: websites/D/Desmos/presence.ts [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+  
+  var presenceData: PresenceData = {
+	largeImageKey:
+	  "logo",
+	startTimestamp: startTime
+  };
```
Comment (EncryptedDev) [inline websites/D/Desmos/presence.ts:72]: Move this to the top of the UpdateData event for convention across the repo
```suggestion
  const presenceData: PresenceData = {
	largeImageKey:
	  "logo",
	startTimestamp: startTime
  };
```

File: websites/D/Desmos/presence.ts [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+  
+  if (graphing === 2) {
+	presenceData.smallImageKey = "logo"
```
Comment (EncryptedDev) [inline websites/D/Desmos/presence.ts:75]: As you've defined this in the presenceData object, you do not have to again
```suggestion
```

File: websites/D/Desmos/presence.ts [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+  	
+  } else 
+  	graphing = 0;
+  
+  
+  const presenceData: PresenceData = {
+    largeImageKey:
+      "logo",
+    startTimestamp: startTime,
+  };
```
Comment (EncryptedDev) [inline websites/D/Desmos/presence.ts:119]: Deja vu, again...

File: websites/D/Desmos/presence.ts [hunk 20]
```diff
@@ -0,0 +1,167 @@
...
+  	numEquations = 0;
+  } else if (["scientific", "fourfunction", "matrix", "practice"].includes(urlPage)) {
+  	graphing = 1;
+  	if (pageType == "Scientific" || pageType == "Fourfunction") 
+  		numEquations = document.getElementsByClassName("dcg-basic-list")[0].childElementCount;
+  	 else if (pageType == "Matrix") 
+  		numEquations = document.getElementsByClassName("dcg-matrix-list")[0].childElementCount;
+  	 else 
+  		numEquations = 0;
+
```
Comment (EncryptedDev) [inline websites/D/Desmos/presence.ts:110]: Remove empty line

File: websites/D/Desmos/presence.ts [hunk 21]
```diff
@@ -0,0 +1,167 @@
...
+  		numEquations = document.getElementsByClassName("dcg-basic-list")[0].childElementCount;
+  	 else if (pageType == "Matrix") 
+  		numEquations = document.getElementsByClassName("dcg-matrix-list")[0].childElementCount;
+  	 else 
+  		numEquations = 0;
+  	
+  } else 
+  	graphing = 0;
+  
+
```
Comment (EncryptedDev) [inline websites/D/Desmos/presence.ts:114]: Remove empty line

File: websites/D/Desmos/presence.ts [hunk 22]
```diff
@@ -0,0 +1,167 @@
...
+  
+  
+  const presenceData: PresenceData = {
+    largeImageKey:
+      "logo",
+    startTimestamp: startTime,
+  };
+  
+  if (graphing == 2) {
+	presenceData.smallImageKey = "logo";
```
Comment (EncryptedDev) [inline websites/D/Desmos/presence.ts:122]: As you've defined this in the presenceData object, you do not have to again

File: websites/D/Desmos/presence.ts [hunk 23]
```diff
@@ -0,0 +1,167 @@
...
+			if (numEquations != 1) presenceData.state += "s"; 
+		} else 
+			delete presenceData.state;
+		
+		if (title != "") 
+			presenceData.details += ": ".concat(title);
+		
+	} else 
+		presenceData.details = "Reading ".concat(pageType);
+
```
Comment (EncryptedDev) [inline websites/D/Desmos/presence.ts:144]: Remove empty line

File: websites/D/Desmos/presence.ts [hunk 24]
```diff
@@ -0,0 +1,167 @@
...
+			delete presenceData.state;
+		
+		if (title != "") 
+			presenceData.details += ": ".concat(title);
+		
+	} else 
+		presenceData.details = "Reading ".concat(pageType);
+	
+  }
+
```
Comment (EncryptedDev) [inline websites/D/Desmos/presence.ts:146]: Remove empty line

File: websites/D/Desmos/presence.ts [hunk 25]
```diff
@@ -0,0 +1,167 @@
...
+  }
+  
+  if (presenceData.details === null) {
+	//This will fire if you do not set presence details
+	desmosPresence.setTrayTitle(); //Clears the tray title for mac users
+	desmosPresence.setActivity(); /*Update the presence with no data, therefore clearing it and making the large image the Discord Application icon, and the text the Discord Application name*/
+  } else {
+	//This will fire if you set presence details
+	desmosPresence.setActivity(presenceData); //Update the presence with all the values from the presenceData object
+  }
```
Comment (EncryptedDev) [inline websites/D/Desmos/presence.ts:154]:
```suggestion
	desmosPresence.setActivity(presenceData);
```

**Review by EncryptedDev — CHANGES_REQUESTED** [review]
I can clearly tell you're new to programming and this is very low quality code. Look at my review, fix the things I've pointed out and there is more to look at, just getting too messy and not understandable with so many comments on your code.

_God knows how this passed Codacy, DeepScan and CodeFactor_

**Comment by Anshroid** [comment]
I think that CodeFactor editing the file itself confused me, so I tried to edit the changes myself as well, and then I ended up with duplicates of everything

File: websites/D/Desmos/presence.ts [hunk 26]
```diff
@@ -0,0 +1,83 @@
...
+        presenceData.details = "Reading ".concat(pageType);
+    
+  }
+  
+  if (presenceData.details === null) {
+    presence.setTrayTitle();
+    presence.setActivity();
+  } else 
+    presence.setActivity(presenceData);
+
```
Comment (EncryptedDev) [inline websites/D/Desmos/presence.ts:82]:
```suggestion
    presence.setActivity(presenceData);
```

File: websites/D/Desmos/presence.ts [hunk 27]
```diff
@@ -0,0 +1,83 @@
...
+            if (numEquations !== 1) presenceData.state += "s"; 
+        } else 
+            delete presenceData.state;
+        
+        if (title !== "") 
+            presenceData.details += ": ".concat(title);
+        
+    } else 
+        presenceData.details = "Reading ".concat(pageType);
+
```
Comment (EncryptedDev) [inline websites/D/Desmos/presence.ts:74]: Remove empty line

**Review by EncryptedDev — CHANGES_REQUESTED** [review]

File: websites/D/Desmos/dist/metadata.json [hunk 28]
```diff
@@ -0,0 +1,22 @@
...
+  "$schema": "https://schemas.premid.app/metadata/1.4",
+  "service": "Desmos",
+  "author": {
+    "name": "Anshroid",
+    "id": "434312233023963137"
+  },
+  "category": "other",
+  "logo": "https://i.imgur.com/MVpfVss.jpg",
+  "thumbnail": "https://i.imgur.com/FkLdvtA.png",
+  "version": "1.0.1",
```
Comment (EncryptedDev) [inline websites/D/Desmos/dist/metadata.json:11]:
```suggestion
  "version": "1.0.0",
```

File: websites/D/Desmos/presence.ts [hunk 29]
```diff
@@ -0,0 +1,75 @@
...
+    if (pageType === "Scientific" || pageType === "Fourfunction") { // These three use a different container for equations
+      numEquations = document.getElementsByClassName("dcg-basic-list")[0].childElementCount;
+    } else if (pageType === "Matrix") 
+      numEquations = document.getElementsByClassName("dcg-matrix-list")[0].childElementCount;
+    else 
+      numEquations = 0;
+  } else 
+    graphing = 0;
+  
+
```
Comment (EncryptedDev) [inline websites/D/Desmos/presence.ts:43]: empty line

File: websites/D/Desmos/presence.ts [hunk 30]
```diff
@@ -0,0 +1,75 @@
...
+      numEquations = 0;
+  } else 
+    graphing = 0;
+  
+  
+  // Setting Presence
+  const presenceData: PresenceData = {
+    largeImageKey: "logo",
+    startTimestamp: startTime
+  };
```
Comment (EncryptedDev) [inline websites/D/Desmos/presence.ts:48]: move this to the top of the UpdateData event

**Review by EncryptedDev — CHANGES_REQUESTED** [review]

**Review by Bas950 — APPROVED** [review]

**Review by Slowlife01 — APPROVED** [review]

**Pull request merged**